### PR TITLE
Adding implicit constraints in Optimizer 

### PR DESCRIPTION
**Pull request by JiaZhou-PU**
--------
Pull Request Description
--------
##### What issue does this change request address? (Use "#" before the issue to link it, i.e., #42.)

Partly address #1223 

##### What are the significant changes in functionality due to this change request?

-Add the rejection flag based on implicit constraints, user needs to assign implicit constraint function in `<functions>` with the response variable included. 
-Change optimizer test function constraints both explicit and implicit.
-Update the user manual for `<Constraint>` and  `<ImplicitConstraint>` with correct parameters.
-Regold function constraints test file

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code.
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [x] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [x] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [x] 8. If an analytic test is changed/added is the the analytic documentation updated/added?
- [x] 9. If any test used as a basis for documentation examples (currently found in `raven/tests/framework/user_guide` and `raven/docs/workshop`) have been changed, the associated documentation must be reviewed and assured the text matches the example.

### CODE REVIEW

**Comment by JiaZhou-PU** [comment]
@alfoa Ready for review

File: doc/user_manual/generated/optimizer.tex [hunk 1]
```diff
@@ -323,14 +323,28 @@ \subsection{GradientDescent}
       the input space of the Model. From a practical point of view, this XML node must contain
       the name of a function defined in the \xmlNode{Functions} block (see
       Section~\ref{sec:functions}).               This external function must contain a method
-      called ``constrain'', which returns 1 for               inputs satisfying the constraints and
-      0 otherwise.
+      called ``constrain'', which returns True for               inputs satisfying the explicit
+      constraints and Flase otherwise.
```
Comment (alfoa) [inline doc/user_manual/generated/optimizer.tex:327]: False

File: doc/user_manual/generated/optimizer.tex [hunk 2]
```diff
@@ -688,14 +702,28 @@ \subsection{SimulatedAnnealing}
...
```
Comment (alfoa) [inline doc/user_manual/generated/optimizer.tex:706]: false

File: framework/Optimizers/Optimizer.py [hunk 3]
```diff
@@ -109,14 +109,36 @@ def getInputSpecification(cls):
...
-    specs.addSub(InputData.assemblyInputFactory('Constraint', contentType=InputTypes.StringType, strictMode=True,
+    ConstraintInput = InputData.parameterInputFactory('Constraint', contentType=InputTypes.StringType, strictMode=True,
         printPriority=150,
         descr=r"""name of \xmlNode{Function} which contains explicit constraints for the sampling of
               the input space of the Model. From a practical point of view, this XML node must contain
               the name of a function defined in the \xmlNode{Functions} block (see Section~\ref{sec:functions}).
-              This external function must contain a method called ``constrain'', which returns 1 for
-              inputs satisfying the constraints and 0 otherwise."""))
+              This external function must contain a method called ``constrain'', which returns True for
+              inputs satisfying the explicit constraints and Flase otherwise.""")
```
Comment (alfoa) [inline framework/Optimizers/Optimizer.py:118]: typo

File: framework/Optimizers/RavenSampled.py [hunk 4]
```diff
@@ -410,6 +411,37 @@ def _applyFunctionalConstraints(self, suggested, previous):
       @ Out, point, dict, adjusted variables
       @ Out, modded, bool, whether point was modified or not
     """
+
+  def _handleImplicitConstraints(self, previous):
+    """
+      Considers all implicit constraints
+      @ In, previous, dict, NORMALIZED previous opt point
+      @ Out, accept, bool, whether point was satisfy implicit constraints
```
Comment (alfoa) [inline framework/Optimizers/RavenSampled.py:419]: satisfied

File: tests/framework/AnalyticModels/optimizing/offset_parabola.py [hunk 5]
```diff
@@ -49,3 +49,16 @@ def constrain(raven):
   if 0.25 < x < 0.75 and 0 < y < 1:
     return False
   return True
+
+def implicitConstrain(raven):
+  """
+    Implicit constrain.
+    @ In, raven, object, raven self
+    @ Out, implicit constrain, point ok or not?
```
Comment (alfoa) [inline tests/framework/AnalyticModels/optimizing/offset_parabola.py:57]: the standard requires, if not a local variable is created (like in this case), to specify in the docstrings the name of the function itself :

```
  @ Out, implicitConstrain, point ok or not?
```

**Comment by JiaZhou-PU** [comment]
@alfoa comments addressed.

**Comment by alfoa** [comment]
failed?

**Comment by JiaZhou-PU** [comment]
> failed?

Invalidated and fail again... on set python environment.

**Comment by JiaZhou-PU** [comment]
@alfoa ready for reviewed again

**Pull request merged**